File: tools/scoring/Score.cpp
#include "Score.hpp"

#include <set>
#include <numeric>
#include <cmath>
#include <iostream>
#include <unordered_map>


class StructureScore {
protected:
    std::vector<int> equ_blocks;
    std::vector<int> var_blocks;
    double score;
    bool scorecomputed;

public:
    StructureScore(const std::vector<int>& equ_blocks, const std::vector<int>& var_blocks)
        : equ_blocks(equ_blocks), var_blocks(var_blocks), score(0.0), scorecomputed(false) {}

    virtual ~StructureScore() = default;

    double get_score() {
        if (!scorecomputed) {
            score = _compute_score();
            scorecomputed = true;
        }
        return score;
    }

protected:
    virtual double _compute_score() = 0;
};

class WhiteScore : public StructureScore {
public:
    WhiteScore(const std::vector<int>& equ_blocks, const std::vector<int>& var_blocks)
        : StructureScore(equ_blocks, var_blocks) {}

protected:
    double _compute_score() override {
        // Get sort indices for equations and variables
        std::vector<size_t> equ_sorted(equ_blocks.size());
        std::iota(equ_sorted.begin(), equ_sorted.end(), 0);
        std::stable_sort(equ_sorted.begin(), equ_sorted.end(),
            [this](size_t i, size_t j) { return equ_blocks[i] < equ_blocks[j]; });

        std::vector<size_t> var_sorted(var_blocks.size());
        std::iota(var_sorted.begin(), var_sorted.end(), 0);
        std::stable_sort(var_sorted.begin(), var_sorted.end(),
            [this](size_t i, size_t j) { return var_blocks[i] < var_blocks[j]; });

        // Create sorted arrays
        std::vector<int> equ_stage(equ_blocks.size());
        std::vector<int> var_stage(var_blocks.size());
        for (size_t i = 0; i < equ_sorted.size(); ++i) {
            equ_stage[i] = equ_blocks[equ_sorted[i]];
        }
        for (size_t i = 0; i < var_sorted.size(); ++i) {
            var_stage[i] = var_blocks[var_sorted[i]];
        }

        // Compute matrix area
        size_t matrix_area = equ_stage.size() * var_stage.size();
        if (matrix_area == 0) {
            return 1.0;
        }

        // Get unique blocks
        std::set<int> blocks_set;
        for (int val : equ_stage) blocks_set.insert(val);
        for (int val : var_stage) blocks_set.insert(val);
        std::vector<int> blocks(blocks_set.begin(), blocks_set.end());

        int max_block = *std::max_element(blocks.begin(), blocks.end());

        size_t block_area = 0;

        for (int block : blocks) {
            if (block == 1) {
                // Linking variables
                size_t var_count = std::count(var_stage.begin(), var_stage.end(), 1);
                size_t equ_count = std::count_if(equ_stage.begin(), equ_stage.end(),
                    [max_block](int val) { return val < max_block; });
                block_area += var_count * equ_count;
            } else if (block == max_block) {
                // Linking constraints
                size_t var_count = var_stage.size();
                size_t equ_count = std::count(equ_stage.begin(), equ_stage.end(), max_block);
                block_area += var_count * equ_count;
            } else {
                // Regular blocks
                size_t var_count = std::count(var_stage.begin(), var_stage.end(), block);
                size_t equ_count = std::count(equ_stage.begin(), equ_stage.end(), block);
                block_area += var_count * equ_count;
            }
        }

        return 1.0 - static_cast<double>(block_area) / static_cast<double>(matrix_area);
    }
};

// Forward declarations for model structures
struct ModelDump {
    std::map<std::string, std::vector<int>> records;
    std::map<std::string, std::vector<std::vector<std::string>>> A_records;
};

class PipsScore : public StructureScore {
private:
    ModelDump model_dump;
    WhiteScore whitescorer;
    int num_blocks;

public:
    PipsScore(const ModelDump& model_dump, const std::vector<int>& equ_blocks, 
              const std::vector<int>& var_blocks)
        : StructureScore(equ_blocks, var_blocks),
          model_dump(model_dump),
          whitescorer(equ_blocks, var_blocks),
          num_blocks(std::max(*std::max_element(equ_blocks.begin(), equ_blocks.end()),
                              *std::max_element(var_blocks.begin(), var_blocks.end()))) {}

protected:
    double _compute_score() override {
        double whitescore = whitescorer.get_score();

        auto [linking_equ_score, linking_equ_frac] = _linking_score(INPUTBLOCKS::EQUATIONS);
        auto [linking_var_score, linking_var_frac] = _linking_score(INPUTBLOCKS::VARIABLES);

        std::cout << "PIPS Score: " << whitescore << " " << linking_equ_score << " "
                  << linking_equ_frac << " " << linking_var_score << " "
                  << linking_var_frac << std::endl;

        return 0.85 * whitescore + 2 * linking_equ_score - 90 * linking_equ_frac
               - 3 * linking_var_score - 90 * linking_var_frac;
    }

    std::pair<double, double> _linking_score(INPUTBLOCKS linking_block_type) {
        const std::vector<int>& linking_labels = 
            (linking_block_type == INPUTBLOCKS::EQUATIONS) ? equ_blocks : var_blocks;
        const std::vector<int>& target_labels = 
            (linking_block_type == INPUTBLOCKS::EQUATIONS) ? var_blocks : equ_blocks;

        std::string linking_type = 
            (linking_block_type == INPUTBLOCKS::EQUATIONS) ? "i" : "j";
        std::string target_type = 
            (linking_block_type == INPUTBLOCKS::EQUATIONS) ? "j" : "i";

        int label = (linking_block_type == INPUTBLOCKS::EQUATIONS) ? num_blocks : 1;

        // Build elem_to_block mapping
        std::unordered_map<std::string, int> elem_to_block;
        const auto& target_records = model_dump.records.at(target_type);
        for (size_t i = 0; i < target_records.size(); ++i) {
            elem_to_block[std::to_string(target_records[i])] = target_labels[i];
        }

        // Process A matrix grouping (simplified - actual implementation depends on data structure)
        const auto& A_records = model_dump.A_records.at(linking_type);
        
        std::vector<std::vector<int>> linking_target_labels;
        for (size_t i = 0; i < linking_labels.size(); ++i) {
            if (linking_labels[i] == label && i < A_records.size()) {
                std::vector<int> target_blocks;
                for (const auto& elem : A_records[i]) {
                    if (elem_to_block.count(elem)) {
                        target_blocks.push_back(elem_to_block[elem]);
                    }
                }
                linking_target_labels.push_back(target_blocks);
            } else {
                linking_target_labels.push_back({});
            }
        }

        auto num_linking = [](const std::vector<int>& target_blocks) -> int {
            std::set<int> unique_blocks(target_blocks.begin(), target_blocks.end());
            if (unique_blocks.size() == 1 && *unique_blocks.begin() == 1) {
                return 10;
            }
            return unique_blocks.size();
        };

        std::vector<double> linking_count;
        for (const auto& target_blocks : linking_target_labels) {
            linking_count.push_back(target_blocks.empty() ? 0.0 : num_linking(target_blocks));
        }

        size_t total_linking = std::count_if(linking_count.begin(), linking_count.end(),
            [](double val) { return val != 0.0; });

        if (total_linking == 0) {
            return {1.0, 0.0};
        }

        size_t three_plus_linking = std::count_if(linking_count.begin(), linking_count.end(),
            [](double val) { return val >= 3.0; });

        return {
            1.0 - static_cast<double>(three_plus_linking) / static_cast<double>(total_linking),
            static_cast<double>(total_linking) / static_cast<double>(linking_labels.size())
        };
    }
};

File: src/utils/anops/score.py
import numpy as np

from .defs import INPUTBLOCKS


class StructureScore:
    """
    a base class for the structure scores
    """

    def __init__(self, equ_blocks, var_blocks):
        self.equ_blocks = equ_blocks
        self.var_blocks = var_blocks

        self.score = 0
        self.scorecomputed = False

    def get_score(self):
        """
        returns the score for the structure. If the score has not be computed,
        then it will be computed before returned

        Returns
        -------
        float
            the score for the structure
        """
        if not self.scorecomputed:
            self.score = self._compute_score()
            self.scorecomputed = True

        return self.score

    def _compute_score(self):
        """
        public function to compute the for the structure

        Returns
        -------
        float
            a score for the structure
        """
        pass


class WhiteScore(StructureScore):
    """
    computes a score of the amount of white space in the matrix area. The white
    space is that which is not the border or a block.
    """

    def __init__(self, equ_blocks, var_blocks):
        super().__init__(equ_blocks, var_blocks)

    def _compute_score(self):
        """
        public function to compute the for the structure

        Returns
        -------
        float
            a score for the structure
        """
        # getting the sort indices for the equations and variables
        equ_sorted = self.equ_blocks.argsort(kind="mergesort")
        var_sorted = self.var_blocks.argsort(kind="mergesort")

        # getting new array's for the equation and variable blocks that
        # correspond to the sorted lists
        equ_stage = self.equ_blocks[equ_sorted]
        var_stage = self.var_blocks[var_sorted]

        # we need to compute matrix area and the block area
        matrix_area = len(equ_stage) * len(var_stage)

        # if there are no variables or no constraints, then the score is 1
        if matrix_area == 0:
            return 1

        block_area = 0

        blocks = np.union1d(np.unique(equ_stage), np.unique(var_stage))

        for block in blocks:
            if block == 1:
                # the linking variables
                block_area += np.count_nonzero(
                    var_stage == 1
                ) * np.count_nonzero(equ_stage < blocks.max())
            elif block == blocks.max():
                # the linking constraints
                block_area += np.count_nonzero(var_stage) * np.count_nonzero(
                    equ_stage == blocks.max()
                )
            else:
                # the blocks
                block_area += np.count_nonzero(
                    var_stage == block
                ) * np.count_nonzero(equ_stage == block)

        return 1 - float(block_area) / float(matrix_area)


class PipsScore(StructureScore):
    """
    a score that finds structures that are suitable for PIPS. Specifically,
    this score computes the white space for the structure and penalises linking
    constraints/variables that link 3 or more blocks.
    """

    def __init__(self, model_dump, equ_blocks, var_blocks):
        super().__init__(equ_blocks, var_blocks)
        self.model_dump = model_dump
        self.whitescorer = WhiteScore(equ_blocks, var_blocks)

        self.num_blocks = int(
            max(np.max(self.equ_blocks), np.max(self.var_blocks))
        )

    def _compute_score(self):
        """
        public function to compute the for the structure

        Returns
        -------
        float
            a score for the structure
        """
        # getting the white score for the structure
        whitescore = self.whitescorer.get_score()

        # computing the linking equations score
        linking_equ_score, linking_equ_frac = self._linking_score(
            INPUTBLOCKS.EQUATIONS
        )

        # computing the linking equations score
        linking_var_score, linking_var_frac = self._linking_score(
            INPUTBLOCKS.VARIABLES
        )

        print(
            "PIPS Score:",
            whitescore,
            linking_equ_score,
            linking_equ_frac,
            linking_var_score,
            linking_var_frac,
        )

        # returning a linear regression based combination of the white score
        # and the linking scores
        return (
            0.85 * whitescore
            + 2 * linking_equ_score
            + -90 * linking_equ_frac
            + -3 * linking_var_score
            + -90 * linking_var_frac
        )

    def _linking_score(self, linking_block_type):
        """
        computes the score for the linking blocks. This is proportional to the
        percentage of the linking region that has variables/constraints that
        link 3 or more blocks

        Returns
        -------
        float
            a score for the linking region
        """
        linking_labels = (
            self.equ_blocks
            if linking_block_type == INPUTBLOCKS.EQUATIONS
            else self.var_blocks
        )
        target_labels = (
            self.var_blocks
            if linking_block_type == INPUTBLOCKS.EQUATIONS
            else self.equ_blocks
        )

        linking_type = (
            "i" if linking_block_type == INPUTBLOCKS.EQUATIONS else "j"
        )
        target_type = (
            "j" if linking_block_type == INPUTBLOCKS.EQUATIONS else "i"
        )

        label = (
            self.num_blocks
            if linking_block_type == INPUTBLOCKS.EQUATIONS
            else 1
        )

        elem_to_block = {
            i: j
            for i, j in zip(
                self.model_dump[target_type].records["uni"], target_labels
            )
        }

        # Get A matrix
        linking_grouping = (
            self.model_dump["A"]
            .records.groupby(linking_type, observed=True)[target_type]
            .apply(list)
            .values
        )

        linking_target_labels = [
            [elem_to_block.get(e) for e in elems]
            if linking_labels[i] == label
            else None
            for i, elems in enumerate(linking_grouping)
        ]

        assert len(linking_labels) == len(linking_grouping)

        def num_linking(target_blocks):
            unique_blocks = np.unique(target_blocks)
            if len(unique_blocks) == 1 and unique_blocks[0] == 1:
                return 10
            else:
                return len(unique_blocks)

        # Get equation annotation
        linking_count = np.array(
            [
                num_linking(target_blocks) if target_blocks is not None else 0
                for i, target_blocks in enumerate(linking_target_labels)
            ]
        ).astype(float)

        total_linking = np.count_nonzero(linking_count)
        # if there are no linking variables/constraints, then there the score
        # is 1
        if total_linking == 0:
            return (1, 0)

        three_plus_linking = np.count_nonzero(linking_count >= 3)
        # import pdb; pdb.set_trace()

        return (
            1 - float(three_plus_linking) / float(total_linking),
            float(total_linking) / float(len(linking_labels)),
        )
